### PR TITLE
Use systemd to reschedule update-status hooks

### PR DESCRIPTION
**Pull request by addyess**
## Overview
Uses systemd to schedule an update-status hook under certain situations to complete node bootstrapping.

## Details
In some circumstances, the charm doesn't complete deployment by the time the last juju event hook fires.  Rather than using event.defer on the last hook event, this creates a reschedulable systemd call to `update-status` hook until the node is ready.

Resolves issue #90

### CODE REVIEW

File: charms/worker/k8s/src/reschedule.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+# Copyright 2024 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (addyess) [reply charms/worker/k8s/src/reschedule.py:1]: i plan on making this a library.  Perhaps the github self hosted runner can adopt it as well!

File: charms/worker/k8s/src/reschedule.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+[Unit]
+Description=Dispatch the {event} event on {unit}
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/timeout {timeout} /usr/bin/bash -c '/usr/bin/juju-exec "{unit}" "JUJU_DISPATCH_PATH={event} ./dispatch"'
+
+[Install]
+WantedBy=multi-user.target
+"""
```
Comment (addyess) [inline charms/worker/k8s/src/reschedule.py:27]: Creates a systemd oneshot service to invoke a juju event

File: charms/worker/k8s/src/reschedule.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+Requires={app}.{event}.service
+
+[Timer]
+Unit={app}.{event}.service
+OnUnitInactiveSec={interval}m
+RandomizedDelaySec={random_delay}m
+
+[Install]
+WantedBy=timers.target
+"""
```
Comment (addyess) [inline charms/worker/k8s/src/reschedule.py:40]: template for a systemd timer which is dependent on the above one-shot service

File: charms/worker/k8s/src/reschedule.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    Args:
+        args (List[str]): arguments to subprocess
+        check_exit (bool): True if we raise on errors
+
+    Returns:
+        stdout, int : tuple to handled output.
+    """
+    s = subprocess.run(args, capture_output=True, check=check_exit)
+    return s.stdout.decode(), s.returncode
```
Comment (addyess) [inline charms/worker/k8s/src/reschedule.py:54]: subprocess wrapper

File: charms/worker/k8s/src/reschedule.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+class TimerEnableError(TimerError):
+    """Raised when unable to enable a event timer."""
+
+
+class TimerDisableError(TimerError):
+    """Raised when unable to disable a event timer."""
+
+
+class TimerStatusError(TimerError):
+    """Raised when unable to check status of a event timer."""
```
Comment (addyess) [inline charms/worker/k8s/src/reschedule.py:70]: Various errors which could be raised by this library

**Review by bschimke95 — COMMENTED** [review]
Great work, added a comment for clarification.

File: charms/worker/k8s/src/charm.py [hunk 6]
```diff
@@ -603,9 +604,12 @@ def _update_status(self):
             status.add(ops.WaitingStatus("Node not Clustered"))
             return
 
+        trigger = reschedule.PeriodicEvent(self)
```
[the remark this answers is not in the document]
Comment (addyess) [reply charms/worker/k8s/src/charm.py:607]: Well, we want to make sure we cancel the `PeriodicEvent` if we happen to handle an `update-status` hook which wasn't triggered by the periodic.  

Maybe the name `PeriodicEvent` is just poor choice.  


the idea i was shooting for was:
- check the status of the cluster (its irrelevant the source of this event -- systemd or jujud)
- if the cluster is not ready, 
    - try to  schedule a re-test in 30 seconds 
    - one may already be scheduled, don't reschedule an existing one
- if the cluster is ready
    -  cancel any retries


that's why you see the creation of this `trigger` object that is either "created" or "cancelled"

[the remark this answers is not in the document]
Comment (bschimke95) [reply charms/worker/k8s/src/charm.py:607]: I see, thanks for the context!

File: charms/worker/k8s/src/reschedule.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bschimke95) [reply charms/worker/k8s/src/reschedule.py:1]: Do you want to do this as part of this PR?

**Review by bschimke95 — APPROVED** [review]
LGTM

File: charms/worker/k8s/src/reschedule.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (addyess) [reply charms/worker/k8s/src/reschedule.py:1]: OH... make a library?  Uhm... I think i'll write an issue and save it as a TODO

**Review by mateoflorido — APPROVED** [review]
Great work, +2

**Pull request merged**